### PR TITLE
Loot tracker grouping loot by NPC/Event

### PR DESCRIPTION
**Pull request by psikoi**
This introduces a new view mode: Grouping by NPC/Event

- Reversible view mode, meaning you can go back to single loots, or, go from single to grouped.
- Adds kill count indicator to grouped logs
- Fixes a small ui margin that was annoying me.
- Adds detailed view for npc/event

peek 2018-09-12 16-04

### CODE REVIEW

**Comment by DannysPVM** [comment]
That looks fantastic @psikoi, keep up the great work man.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 1]
```diff
@@ -160,7 +183,7 @@ public void onPlayerLootReceived(final PlayerLootReceived playerLootReceived)
 		final String name = player.getName();
 		final int combat = player.getCombatLevel();
 		final LootTrackerItemEntry[] entries = buildEntries(stack(items));
-		SwingUtilities.invokeLater(() -> panel.addLog(name, combat, entries));
+		SwingUtilities.invokeLater(() -> panel.addEntry(name, combat, entries, config.groupLoot()));
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: I would just do setGroupLoot on the panel from configChanged instead of adding new method param

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 2]
```diff
@@ -149,7 +172,7 @@ public void onNpcLootReceived(final NpcLootReceived npcLootReceived)
 		final String name = npc.getName();
 		final int combat = npc.getCombatLevel();
 		final LootTrackerItemEntry[] entries = buildEntries(stack(items));
-		SwingUtilities.invokeLater(() -> panel.addLog(name, combat, entries));
+		SwingUtilities.invokeLater(() -> panel.addEntry(name, combat, entries, config.groupLoot()));
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: See my next comment

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 3]
```diff
@@ -203,7 +226,7 @@ public void onWidgetLoaded(WidgetLoaded event)
 		if (!items.isEmpty())
 		{
 			final LootTrackerItemEntry[] entries = buildEntries(stack(items));
-			SwingUtilities.invokeLater(() -> panel.addLog(eventType, -1, entries));
+			SwingUtilities.invokeLater(() -> panel.addEntry(eventType, -1, entries, config.groupLoot()));
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: See my previous comment

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java [hunk 4]
```diff
@@ -141,6 +153,17 @@ protected void shutDown()
...
 
+	@Subscribe
+	public void onConfigChanged(ConfigChanged ev)
+	{
+		if (!LootTrackerConfig.GROUP_KEY.equals(ev.getGroup()))
+		{
+			return;
+		}
+
+		panel.rebuildLogs(config.groupLoot());
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: See my comment about config.groupLoot()

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java [hunk 5]
```diff
@@ -131,39 +142,105 @@ private static String htmlLabel(String key, long value)
...
-		overallGp += box.getTotalPrice();
+		overallGp += log.getTotalPrice();
 		overallKills += 1;
 		updateOverall();
 
+		logs.add(log);
+		logsContainer.add(log, 0);
+	}
+
+	void rebuildLogs(boolean groupLoot)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: rename this to refresh, remove method param

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java [hunk 6]
```diff
@@ -40,13 +43,29 @@
 import net.runelite.client.util.StackFormatter;
 
 @Getter
-class LootTrackerBox extends JPanel
+class LootTrackerLog extends JPanel
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: LootTrackerBox was better name as it indicated more that it was component, this is more confusing. Use the original name

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java [hunk 7]
```diff
@@ -40,13 +43,29 @@
...
+	private final JLabel priceLabel = new JLabel();
+	private final JLabel subTitleLabel = new JLabel();
+
+	private final ItemManager itemManager;
+
+	private long totalPrice;
+	private int totalKills;
+
+	@Getter
+	private String title;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: Just override equals and compare 2 lootTrackerBoxes instead of exposing title just for that

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: Also, rename this to id or something

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java [hunk 8]
```diff
@@ -102,8 +146,52 @@
 			itemContainer.add(slotContainer);
 		}
 
-		add(logTitle, BorderLayout.NORTH);
-		add(itemContainer, BorderLayout.CENTER);
+		itemContainer.repaint();
+	}
+
+	private List<LootTrackerItemEntry> getCombinedItems()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: Remove this method and just move it to rebuildItems() and calculate total in same loop

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java [hunk 9]
```diff
@@ -102,8 +146,52 @@
...
+				list.add(entry);
+			}
+		}
+
+		list.sort((i1, i2) -> i1.getPrice() < i2.getPrice() ? 1 : -1);
+
+		return list;
+	}
+
+	private List<LootTrackerItemEntry> getAllItems()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: just iterate the entries in the rebuildItems, no need for new list

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java [hunk 10]
```diff
@@ -61,36 +80,61 @@
...
+			priceLabel.setText(StackFormatter.quantityToStackSize(totalPrice) + " gp");
 		}
 
+		if (totalKills > 1)
+		{
+			subTitleLabel.setText("x " + totalKills);
+		}
+	}
+
+	private void rebuildItems()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: rename this to buildItems

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java [hunk 11]
```diff
@@ -61,36 +80,61 @@
...
+
+	void addEntry(LootTrackerEntry entry)
+	{
+		entries.add(entry);
+		totalKills++;
+		totalPrice = calculatePrice();
+
+		rebuildItems();
 
 		if (totalPrice > 0)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: Just always append the price unconditionally to make the look consistent

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java [hunk 12]
```diff
@@ -61,36 +80,61 @@
...
 		{
-			final JLabel priceLabel = new JLabel(StackFormatter.quantityToStackSize(totalPrice) + " gp");
-			priceLabel.setFont(FontManager.getRunescapeSmallFont());
-			priceLabel.setForeground(ColorScheme.LIGHT_GRAY_COLOR);
-			logTitle.add(priceLabel, BorderLayout.EAST);
+			priceLabel.setText(StackFormatter.quantityToStackSize(totalPrice) + " gp");
 		}
 
+		if (totalKills > 1)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: same here, make it x 1

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java [hunk 13]
```diff
@@ -40,13 +43,29 @@
...
+
+	private long totalPrice;
+	private int totalKills;
+
+	@Getter
+	private String title;
+
+	private List<LootTrackerEntry> entries = new ArrayList<>();
+
+	LootTrackerLog(final ItemManager itemManager, LootTrackerEntry entry)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: Having both initial construction from entry and method addEntry seems weird.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java [hunk 14]
```diff
@@ -115,12 +203,12 @@ private String buildToolTip(LootTrackerItemEntry item)
 		return name + " x " + quantity + " (" + StackFormatter.quantityToStackSize(price) + ")";
 	}
 
-	private static long calculatePrice(final LootTrackerItemEntry[] itemStacks)
+	private long calculatePrice()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: see my comment about calculating total

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java [hunk 15]
```diff
@@ -131,39 +142,105 @@ private static String htmlLabel(String key, long value)
 		return String.format(HTML_LABEL_TEMPLATE, ColorUtil.toHexColor(ColorScheme.LIGHT_GRAY_COLOR), key, valueStr);
 	}
 
-	void addLog(final String eventName, final int actorLevel, LootTrackerItemEntry[] items)
+	void addEntry(final String eventName, final int actorLevel, LootTrackerItemEntry[] items, boolean groupLoot)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: Maybe you should just send the LootTrackerEntry as param.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java [hunk 16]
```diff
@@ -131,39 +142,105 @@ private static String htmlLabel(String key, long value)
...
+	{
+		final String subTitle = actorLevel > -1 ? "(lvl-" + actorLevel + ")" : "";
+
+		LootTrackerEntry entry = new LootTrackerEntry(eventName, subTitle, System.currentTimeMillis(), items);
+		entries.add(entry);
+
+		if (!groupLoot || !stackEntry(entry))
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: Use list.contains. Also, this logic is very cryptic. 
I would just make the entries map, and do:
```
existingEntry = map.get(e.getname());
if (entry != null)
{
  existingEntry.combineWith(entry);
}

overallGp = map.values().stream().map(Entry::getTotalPrice()).sum();
```

also probably I would do that by using map.compute or something.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java [hunk 17]
```diff
@@ -131,39 +142,105 @@ private static String htmlLabel(String key, long value)
...
+	void rebuildLogs(boolean groupLoot)
+	{
+		logsContainer.removeAll();
+		logs.clear();
+
+		overallGp = 0;
+		overallKills = 0;
+		updateOverall();
+
+		for (LootTrackerEntry entry : entries)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: see my previous comment, also this logic looks duplicated

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java (hunk 13 above)
Comment (psikoi) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: Because every log contains atleast one entry, the initial one, new ones can be added later

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java (hunk 1 above)
Comment (psikoi) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: I was trying to avoid having that sort of stuff in the UI classes

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPlugin.java]: I dont mean of you adding configChanged even to panel. But to make setter and set it on panel in configChanged event from plugin. I dont see how adding parameter to multiple methods for param that is basically global is better.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java (hunk 5 above)
Comment (psikoi) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: You sure? It is only used to re-construct the logs when the grouping config changes (from single to grouped or vice versa), not to refresh any view, rebuild sounds more appropriate

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java (hunk 7 above)
Comment (psikoi) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: Won't I have to expose the id then? to compare between two boxes

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java (hunk 8 above)
Comment (psikoi) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: But rebuild items is only executed when the config changes, not everytime a new item/log is added, so the price would always be oudated until the config changed

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java (hunk 12 above)
Comment (psikoi) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: Then single logs would not show combat, even if grouping was disabled.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java (hunk 7 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: Nope, you would compare them based on title. `@EqualsAndHashCode(of='title')` or something like that

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java (hunk 5 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: then just rebuild i guess

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java (hunk 7 above)
Comment (psikoi) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: Ah, didn't know about that annotation

Comment (psikoi) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerLog.java]: Still, why would I compare two boxes if I don't need to create a second in the first place, I can just search for that ID, and if a box with that Id already exists, add items to it, instead of creating a new box and comparing it, only to then throw it away

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerConfig.java [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+	String GROUP_KEY = "loottracker";
+
+	@ConfigItem(
+		keyName = "groupLoot",
+		name = "Group loot by NPC/Activity",
+		description = "Groups the loot received from a NPC or Activity into one single log."
+	)
+	default boolean groupLoot()
+	{
+		return true;
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerConfig.java]: This should probably default to the existing behavior?

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerConfig.java]: The group layout is more useful than per-kill layout in almost every scenario, I think it should be default, and a lot of people want just that.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java [hunk 19]
```diff
@@ -199,11 +234,19 @@ private void buildBox(LootTrackerRecord record)
...
-		final JPopupMenu popupMenu = new JPopupMenu();
-		popupMenu.setBorder(new EmptyBorder(5, 5, 5, 5));
 		popupMenu.add(reset);
+
+		if (groupLoot)
+		{
+			// Create details menu
+			final JMenuItem details = new JMenuItem("View details");
+			details.addActionListener(e ->
+			{
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: remove the unnecessary { } around and make this 1 line

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java [hunk 20]
```diff
@@ -199,11 +234,19 @@ private void buildBox(LootTrackerRecord record)
...
+				showDetailedView(box);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: Just set the detailedKey here, instead of exposing the ID from box (here you already have access to ID)

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerBox.java [hunk 21]
```diff
@@ -50,6 +50,8 @@
 	private final JLabel priceLabel = new JLabel();
 	private final JLabel subTitleLabel = new JLabel();
 	private final ItemManager itemManager;
+
+	@Getter
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerBox.java]: Revert this

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java [hunk 22]
```diff
@@ -56,6 +56,7 @@
 
 	// Handle loot boxes
 	private final JPanel logsContainer = new JPanel();
+	private final JPanel layoutPanel = new JPanel();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: Move this back to constructor? It is unused from outside of it

**Comment by Nightfirecat** [comment]
It's a little jarring to have the scrollbar appear and disappear when switching from a view that doesn't take all the vertical space to one which does. Could we have a scrollbar permanently there, and disabled/taking the full height when it's not needed, or some other way to avoid that little jump when switching between groupings?

Also, the grouping buttons don't have tooltips so there's no way to tell exactly what they represent.

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java [hunk 23]
```diff
@@ -143,8 +169,13 @@ void add(final String eventName, final int actorLevel, LootTrackerItem[] items)
 	{
 		final String subTitle = actorLevel > -1 ? "(lvl-" + actorLevel + ")" : "";
 		final LootTrackerRecord record = new LootTrackerRecord(eventName, subTitle, items);
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: Any reason for this empty line?

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java [hunk 24]
```diff
@@ -189,6 +220,10 @@ private void buildBox(LootTrackerRecord record)
 		final LootTrackerBox box = new LootTrackerBox(itemManager, record.getTitle(), record.getSubTitle());
 		box.combine(record);
 
+		// Create popup menu
+		final JPopupMenu popupMenu = new JPopupMenu();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: Why was this moved here?

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerConfig.java [hunk 25]
```diff
@@ -36,10 +36,19 @@
 	@ConfigItem(
 		keyName = "groupLoot",
 		name = "Group loot by NPC/Activity",
-		description = "Groups the loot received from a NPC or Activity into one single log."
+		description = "Groups the loot received from a NPC or Activity into one single log.",
+		hidden = true
 	)
 	default boolean groupLoot()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerConfig.java]: change this to get, and description for the field is unnecessary anymore

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java [hunk 26]
```diff
@@ -64,18 +76,39 @@
 	private final JLabel overallGpLabel = new JLabel();
 	private final JLabel overallIcon = new JLabel();
 
+	private final JPanel actionsContainer = new JPanel();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: Are you sure all those need to be instance fields? And not created only where they are used?

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerRecord.java [hunk 27]
```diff
@@ -24,6 +24,7 @@
  */
 package net.runelite.client.plugins.loottracker;
 
+import java.util.ArrayList;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerRecord.java]: Unused import

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java [hunk 28]
```diff
@@ -257,6 +358,15 @@ private void buildBox(LootTrackerRecord record)
 		updateOverall();
 	}
 
+	void onGroupingModeChanged(boolean group)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: on? rename this to changeGroupingMode, onX/Y would mean this is some sort of event, but it is not

File: runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java [hunk 29]
```diff
@@ -281,11 +282,12 @@ void add(final String eventName, final int actorLevel, LootTrackerItem[] items)
 	/**
 	 * Rebuilds all the boxes from scratch using existing listed records, depending on the grouping mode.
 	 */
-	void rebuild()
+	private void rebuild()
 	{
 		logsContainer.removeAll();
 		boxes.clear();
-		records.stream().filter(r -> detailedKey == null || r.getTitle().equals(detailedKey)).forEach(this::buildBox);
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/loottracker/LootTrackerPanel.java]: any reason for the new line

**Comment by deathbeam** [comment]
Added tooltips
